src/Core/Util/Url/BackUrlProvider.php

```diff
@@ -0,0 +1,68 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Util\Url;
+
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * gets back url.
+ */
+class BackUrlProvider
+{
+    /**
+     * @var RequestStack|null
+     */
+    private $requestStack;
+
+    /**
+     * @param RequestStack|null $requestStack
+     */
+    public function __construct($requestStack)
```

Comment:
Let's avoid injecting the RequestStack in services if we can 😉and here we can ! (see https://igor.io/2013/03/31/stateless-services.html for detailed explanation about this matter)

You can pass the request in `public function getBackUrl(Request $request)`.

2 side-effects that are great benefits (apart from the benefit number one which is "do not depend on request stack"):
- we can use this service in CLI !
- we can use this service on multiple Requests to retrieve the back parameter, not only the one provided by RequestStack

Reply:
great article! Indeed it was easy to avoid RequestStack in that service - got reduced its usage and right now it is used in `PathWithBackUrlExtension` twig extension only.